### PR TITLE
  X509_cmp(): Fix comparison in case x509v3_cache_extensions() failed to due to invalid cert

### LINKED ISSUE #13754: d2i_X509 and d2i_X509_AUX behavior has become inconsistent and non-intuitive and is not documented
@slontis wrote in https://github.com/openssl/openssl/pull/12153#issuecomment-650094516:

> NOTE that `d2i_X509()` calls `X509v3_cache_extensions()` if the cert object is passed in.

Why has this non-intuitive behavior been added? At least, this must be documented.
If calling `X509v3_cache_extensions()` is really helpful then better do this also in case no cert object is passed in.

(**Update:** The issue described in this paragraph is fixed by #13755:
The call of  `X509v3_cache_extensions()` causes trouble: if the caller passed in a pointer to a NULL cert pointer the caller must be very careful in case NULL is returned after the cert was parsed successfully but is invalid such that `X509v3_cache_extensions()` failed.
In this case the caller needs to free the cert object which has been allocated regardless.)

I have already fixed part of this in #13755 (because I needed it there) such that `d2i_X509()` behaves consistently at least with `d2i_X509_AUX`, which deallocates the just created cert object in case an error occurred in its further processing. This way a mem leak is avoided.

@slontis, ~do you agree with the small improvement I did, and~ 
can you please properly document `d2i_X509()` and `d2i_X509_AUX()`?
The description in `i2d doc/man3/d2i_X509.pod` so far does not cover that new special feature of conditionally calling `X509v3_cache_extensions()`in `d2i_X509()` and thus also in `d2i_X509_AUX()`.
Moreover, the deallocation behavior for any passed-in object pointer is not documented for any `d2i_TYPE()` , neither for  `d2i_X509_{CERT_,}AUX()`.

### PR DESCRIPTION
**Pull request by DDvO**
This is the upstream fix for #13698 reported for v1.1.1 as discussed there,
by checking more specifically if the hash value computation succeeded.
This required improving the error reporting of `x509v3_cache_extensions()`

Adds a test using the example cert given there, which has invalid SKID and AKID encodings.

Also makes the `d2i_X509()` deallocation behavior consistent with `d2i_X509_AUX()`,
which is needed for the new test and partly fixes #13754.

- [x] tests are added or updated

### CODE REVIEW

File: doc/man3/X509_get_extension_flags.pod [hunk 1]
```diff
@@ -78,12 +78,17 @@ The certificate contains an unhandled critical extension.
...
 
-Some certificate extension values are invalid or inconsistent. The
-certificate should be rejected.
+Some certificate extension values are invalid or inconsistent.
+The certificate should be rejected.
 This bit may also be raised after an out-of-memory error while
 processing the X509 object, so it may not be related to the processed
 ASN1 object itself.
 
+=item B<EXFLAG_HASH_INVALID>
```
Comment (kaduk) [inline doc/man3/X509_get_extension_flags.pod:87]: I wonder if a name like `EXFLAG_NO_FINGERPRINT` might make more sense -- the current name feels like it's trying to be very similar to `EXFLAG_INVALID`, but they seem to reflect qualitatively different things -- a property of the certificate vs our ability to process it.

Comment (DDvO) [reply doc/man3/X509_get_extension_flags.pod]: Good point - done.

**Comment by t8m** [comment]
Could you please rebase and squash the commits?

File: crypto/asn1/charmap.h [hunk 2]
```diff
@@ -2,7 +2,7 @@
  * WARNING: do not edit!
  * Generated by crypto/asn1/charmap.pl
  *
- * Copyright 2000-2020 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2000-2021 The OpenSSL Project Authors. All Rights Reserved.
  *
```
Comment (beldmit) [inline crypto/asn1/charmap.h:6]: Do we really need the copyright year change without any other changes?

Comment (DDvO) [reply crypto/asn1/charmap.h:6]: Yes, unfortunately we do need to manually update the copyright year, due to an issue I've reported earlier today: #13765.
Yet this won't be part of this PR any more as soon as I've rebased this to the (same) workaround meanwhile is merged.

**Comment by DDvO** [comment]
> Could you please rebase and squash the commits?

Sure. done.

File: crypto/x509/x_x509.c [hunk 3]
```diff
@@ -125,12 +125,16 @@ IMPLEMENT_ASN1_DUP_FUNCTION(X509)
 X509 *d2i_X509(X509 **a, const unsigned char **in, long len)
 {
     X509 *cert = NULL;
+    int free_on_error = a != NULL && *a == NULL;
 
     cert = (X509 *)ASN1_item_d2i((ASN1_VALUE **)a, in, len, (X509_it()));
     /* Only cache the extensions if the cert object was passed in */
     if (cert != NULL && a != NULL) {
-        if (!x509v3_cache_extensions(cert))
+        if (!x509v3_cache_extensions(cert)) {
```
Comment (t8m) [inline crypto/x509/x_x509.c:133]: Changes in this file belong to the other commit probably?

Comment (DDvO) [reply crypto/x509/x_x509.c:133]: Oops, good spotting. Fixed.

File: test/certs/invalid-cert.pem [hunk 4]
```diff
@@ -0,0 +1,19 @@
+-----BEGIN TRUSTED CERTIFICATE-----
```
Comment (t8m) [inline test/certs/invalid-cert.pem:1]: Also this addition belongs to the main commit?

Comment (DDvO) [reply test/certs/invalid-cert.pem:1]: Oops; fixed as well.

File: doc/man3/X509_get_extension_flags.pod [hunk 5]
```diff
@@ -78,12 +78,17 @@ The certificate contains an unhandled critical extension.
...
-certificate should be rejected.
+Some certificate extension values are invalid or inconsistent.
+The certificate should be rejected.
 This bit may also be raised after an out-of-memory error while
 processing the X509 object, so it may not be related to the processed
 ASN1 object itself.
 
+=item B<EXFLAG_NO_FINGERPRINT>
+
+Failed to compute the internal SHA1 hash value of the certificiate or CRL.
```
Comment (misery) [inline doc/man3/X509_get_extension_flags.pod:89]: same typo: certificiate

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged - thanks @t8m

**Pull request closed without merging**